### ISSUE TITLE
 BasicClassProxyTestCase.ProxyForNonPublicClass test failure after upgrading to .NET Core RTM

### ISSUE THREAD

**Issue by jeremymeng** [castleproject/Core#184]
```
Assert.Throws() Failure
Expected: typeof(Castle.DynamicProxy.Generators.GeneratorException)
Actual:   typeof(System.ArgumentNullException): Value cannot be null.
Parameter name: classToProxy
   at Castle.DynamicProxy.ProxyGenerator.CreateClassProxy(Type classToProxy, Type[] additionalInterfacesToProxy, ProxyGenerationOptions options, Object[] constructorArguments, IInterceptor[] interceptors) in D:\work\d545f799076bb8eb\src\Castle.Core\DynamicProxy\ProxyGenerator.cs:line 1424
   at Castle.DynamicProxy.ProxyGenerator.CreateClassProxy(Type classToProxy, IInterceptor[] interceptors) in D:\work\d545f799076bb8eb\src\Castle.Core\DynamicProxy\ProxyGenerator.cs:line 1334
   at CastleTests.BasicClassProxyTestCase.<>c__DisplayClass2_0.<ProxyForNonPublicClass>b__0() in D:\work\d545f799076bb8eb\src\Castle.Core.Tests\BasicClassProxyTestCase.cs:line 90
```

**Comment by jeremymeng** [comment]
Most likely because the type `System.AppDomainInitializerInfo` is gone in .NET Core. I changed to use another internal type `System.StubHelpers.StubHelpers`, which gave me the same `GeneratorException` type, but with a slightly different exception message.  The assembly name in the message is `System.Private.CoreLib` instead of `mscorlib`: `...because assembly System.Private.CoreLib is strong-named.`.  On .NET Core 1.0 RTM mscorlib just forwards types to System.Pivate.CoreLib.

Ideally the test shouldn't depend on external things that is subject to change.  However, I could see that adding an assembly to test this scenario seems an overkill. We could change the verification to just `string.StartsWith()`, or a little more involved using Regex.

@jonorossi what's your opinion?

**Comment by kzu** [comment]
For tracking purposes, linking https://github.com/moq/moq4/pull/225 from here.

**Comment by jonorossi** [comment]
@jeremymeng change to using `StartsWith` sounds fine to me.

@kzu isn't this just a failing unit test, how is Moq broken? If you are just needing a new package using the RTM runtime packages, then once this one is fixed I'll publish alpha002 straight away to unblock everyone before the other bits get finished and to allow all the mocking libraries to ensure things all work before we release.

**Comment by kzu** [comment]
We're blocked from shipping a stable version until you ship a stable one
too ;)

**Comment by jonorossi** [comment]
> We're blocked from shipping a stable version until you ship a stable one too ;)

@kzu I'm definitely aware of that :), just wondering why you linked up to this specific issue.

**Comment by kzu** [comment]
No particular reason :)

Just thought you might not ship with failing tests, he he

**Comment by kalahari** [comment]
I submitted #186 based on @jeremymeng's comment above. The builds look good locally.

**Comment by jonorossi** [comment]
As I mentioned in #186, this failed on .NET 3.5.

@jeremymeng `StubHelpers` is a bit obscure, why not something like `System.__Canon` or `System.__ComObject` which are well known internal implementation details?

**Comment by jeremymeng** [comment]
Sorry for the NET35 failure.  I must have missed the error.  I search for an internal class using reference source then used ildasm to open the .NET Core version of `mscorlib` (`C:\Program Files\dotnet\shared\Microsoft.NETCore.App\1.0.0\mscorlib.dll`) to make sure that the type exists.  `System.__Canon` and `System.__ComObject` are not there so they would fail the .NET Core build.

**Comment by jonorossi** [comment]
Is that copy of `mscorlib` a reference assembly or the one that actually gets run? Because `System.__Canon` is definitely there in the coreclr's `mscorlib`:
https://github.com/dotnet/coreclr/blob/840a6ca79d4643cd0cfed9444e5658933d901618/src/mscorlib/src/System/Object.cs#L229

**Comment by jeremymeng** [comment]
It is the runtime.

![image](https://cloud.githubusercontent.com/assets/7583839/16671185/e9c83008-4452-11e6-8667-75381cecae07.png)

on .NET Core, `System.__Canon` is in `System.Private.CoreLib`: https://github.com/dotnet/corert/blob/4376c55dd018d8b7b9fed82449728711231ba266/src/System.Private.CoreLib/src/System/__Canon.cs

`System.__ComObject` seems in `System.Private.Interop`: https://github.com/dotnet/corert/blob/62e48db28c7fdaa8968107eb5e9b66962750671f/src/System.Private.Interop/src/Shared/__ComObject.cs

**Comment by jeremymeng** [comment]
How about `"NLog.Internal.LoggerConfiguration, NLog"`? It should work for all build configurations.

**Comment by jonorossi** [comment]
@jeremymeng you linked to the .NET Native/AOT runtime code not coreclr, however I can see that the coreclr repo does also have a `mscorlib.csproj` for the facade mscorlib to support forwarding types, while also having `System.Private.CoreLib.csproj` for the actual internal implementation that used to live in there. I assume this work was to make corert work easier and separate the mscorlib contract from the implementation since it is different with the System.Runtime based contract.

Anyway to the real point, I didn't really want a reference to NLog in the DP tests so I've merged in #197 using `System.__Canon` without specifying the assembly name. All the builds are now green.

**Comment by jeremymeng** [comment]
Great!
